applications/nrf_desktop/src/hw_interface/click_detector.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#define MODULE click_detector
+#include "module_state_event.h"
+
+#include "button_event.h"
+#include "power_event.h"
+#include "click_event.h"
+
+#include "click_detector_def.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CLICK_DETECTOR_LOG_LEVEL);
+
+#define CLICK_CHECK_PERIOD	K_MSEC(100)
+
+#define SHORT_CLICK_MAX		K_MSEC(500)
+#define LONG_CLICK_MIN		K_MSEC(5000)
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_SLEEP,
+};
+
+struct key_state {
+	bool pressed;
+	int click_short_timeout;
+	size_t pressed_time;
+};
+
+struct key_state keys[ARRAY_SIZE(click_detector_configs)];
+
+static enum state state;
+struct k_delayed_work click_check;
+
+
+static void submit_click_event(u16_t key_id, enum click click)
+{
+	struct click_event *event = new_click_event();
+
+	event->key_id = key_id;
+	event->click = click;
+
+	EVENT_SUBMIT(event);
+}
+
+static enum click get_click(u32_t time_diff)
+{
+	if (time_diff < SHORT_CLICK_MAX) {
+		return CLICK_SHORT;
+	}
+
+	if (time_diff > LONG_CLICK_MIN) {
+		return CLICK_LONG;
+	}
+
+	return CLICK_NONE;
+}
+
+static void update_click_short_timeout(struct key_state *key, u16_t key_id)
+{
+	if (key->click_short_timeout) {
+		key->click_short_timeout -= CLICK_CHECK_PERIOD;
+
+		if (key->click_short_timeout <= 0) {
+			submit_click_event(key_id, CLICK_SHORT);
+			key->click_short_timeout = 0;
+		}
+	}
+}
+
+static void update_click_time(struct key_state *key, u16_t key_id)
+{
+	if (key->pressed) {
+		key->pressed_time += CLICK_CHECK_PERIOD;
+
+		if (get_click(key->pressed_time) == CLICK_LONG) {
+			key->pressed = false;
+			key->pressed_time = 0;
+			submit_click_event(key_id, CLICK_LONG);
+		}
+	}
+}
+
+static void click_check_fn(struct k_work *work)
+{
+	__ASSERT_NO_MSG(state == STATE_ACTIVE);
+
+	for (size_t i = 0; i < ARRAY_SIZE(keys); i++) {
+		update_click_short_timeout(&keys[i],
+					   click_detector_configs[i].key_id);
+		update_click_time(&keys[i], click_detector_configs[i].key_id);
+	}
+
+	k_delayed_work_submit(&click_check, CLICK_CHECK_PERIOD);
+}
+
+static int init(void)
+{
+	BUILD_ASSERT(ARRAY_SIZE(keys) == ARRAY_SIZE(click_detector_configs));
+
+	k_delayed_work_init(&click_check, click_check_fn);
+	k_delayed_work_submit(&click_check, CLICK_CHECK_PERIOD);
+
+	state = STATE_ACTIVE;
+
+	return 0;
```

Comment:
Is it required if function may not fail?

Reply:
ok